src/librbd/image/PreRemoveRequest.cc

```diff
@@ -272,7 +276,10 @@ void PreRemoveRequest<I>::handle_remove_snapshot(int r) {
   auto cct = m_image_ctx->cct;
   ldout(cct, 5) << "r=" << r << dendl;
 
-  if (r < 0 && r != -ENOENT) {
+  if (r == -EBUSY && m_attached_children != nullptr) {
+    ldout(cct, 5) << "skipping attached child" << dendl;
+    *m_attached_children = true;
```

Comment:
Is this just to avoid the error message below? Otherwise, you could just check for ```-ENOTEMPTY```, right?

Reply:
Without this branch the request will terminate below, instead of trying to remove other snapshots if they remain. I think it is a good idea to remove all snapshots we can. And yes, we don't want a error message in this case.

The `attached_children` param for `PreRemoveRequest` serves two purposes:
1) (flag) If it does not null tell the requests we want to continue in a case like this;
2) (out) inform the caller about not removed snapshots.

I don't like this interface much but it allows to avoid adding two separate params. If you think it is ok to return here and potentially leave snapshots that could be removed, then yes, I could just check for `-ENOEMPTY`. Then I think we will need to suppress the error message below for `r == -EBUSY`.

Reply:
Actually -- I think it would be good behavior to always attempt to remove all auto-prune snapshots and to reduce that error to a ```ldout(cct, 5)```-ish log message. Then you can avoid passing a new param and can just use the result code.

Reply:
Ah, their is a complication. `PreRemoveReques` will also return `-ENOEMPTY` if the image has a "normal" (not auto-delete) snapshot. And we need to distinguish this case some how to propagate the error.

Reply:
Perhaps just introduce a new error code for the child case like ```-ECHILD```?